nexus/db-model/src/region.rs

```diff
@@ -30,7 +30,7 @@ pub struct Region {
     #[diesel(embed)]
     identity: RegionIdentity,
 
-    dataset_id: Uuid,
```

Comment:
You wrote

```
        // As disks are decommissioned, this "dataset_id" field is set to NULL
        // so we can quickly identify these regions, even though the underlying
        // datasets + zpools are deleted.
```

but this doesn't work. `dataset_id` is our informal foreign key into the dataset table and this relationship is depended on in a lot of places:

- when constructing the region's IP address (see: https://github.com/oxidecomputer/omicron/blob/5b32a8d8f7a23b87c170c0512e7c39e330bfeb91/nexus/db-queries/src/db/datastore/region.rs#L487). This PR changes that function to return `None` if the region's `dataset_id` is `None`, which can't work: when a disk is expunged, volume reconstruction operations (region and snapshot replacement) start. The dataset record (and the region's association) needs to stick around for those operations. During the region replacement start saga, `srrs_get_old_region_address` uses `region_addr`, and that won't return anything if `dataset_id` is set to None, and in the branch that is meant to handle a `Gone` this PR adds a bail if `dataset_id` is None. This will halt the whole region replacement. The intention was that these reconstruction operations would happen in the background (based on if policies were set to expunged) but with this PR those operations now will race with reconfigurator.

- `get_allocated_regions` is no longer idempotent for a given volume_id: if dataset_id of a region is set to `None`, that region won't be returned by the query anymore. Depending on the calling saga, this could cause further region allocations to meet some redundancy level.

- the same is true for `regions_to_delete` and `find_deleted_volume_regions`, and additionally this will cause those regions (and region records) to never be cleaned up.

There's another problem that's unrelated: if the expunged dataset's IP is reused during the allocation of a new dataset, this could lead to the problem where the Upstairs [does not validate the UUID of downstairs it connects to](https://github.com/oxidecomputer/crucible/issues/726). An upstairs could connect to what it thinks is an old dataset with one of the regions of a region set, when in fact it could be connecting to something totally different. A mismatch in region definition would cause a panic. This happened in a different way [when Nexus deleted a read-only parent, then provisioned a region, and the crucible agent reused the port for that new downstairs](https://github.com/oxidecomputer/crucible/issues/1090). I don't think this PR affects how sled IPs are reused at all, just wanted to bring it up again (maybe the presence of a dataset record (even if it's deleted) would mean that the IP won't be reused? I haven't looked in a while but I don't think so, I don't think this affects sled IP allocation)

I propose undoing this schema change, and creating an issue for the garbage collection of the dataset and zpool records. Feel free to assign it to me if you like and I can follow up after the snapshot replacement work is in.

Reply:
I've pretty significantly updated this PR in https://github.com/oxidecomputer/omicron/pull/6059/commits/8467942be65ca51ed45656a03cc024cfc2f5f15b

To summarize:

- I'm no longer nullifying the `dataset_id` columns of regions
- I'm doing "zpool / dataset" record deletion in a background task, still only on decommissioned disks
- The record deletion is conditional on no regions nor region snapshots existing